src/common/protocol-handler/router.ts

```diff
@@ -0,0 +1,205 @@
+export type FallbackHandler = (name: string) => Promise<boolean>;
+
+export enum HandlerType {
+  INTERNAL = "internal",
+  EXTENSION = "extension",
+}
+
+interface ExtensionParams {
+  handlerType: HandlerType.EXTENSION,
+  extensionId: string,
+}
+
+interface InternalParams {
+  handlerType: HandlerType.INTERNAL,
+}
+
+type BaseParams = (ExtensionParams | InternalParams);
+
+export type RegisterParams = BaseParams & {
+  handlerId: string,
+  pathSchema: string,
+};
+
+export interface DeregisterParams {
+  extensionId: string,
+}
+
+export type BackChannelParams = BaseParams & {
+  params: RouteParams;
+  handlerId: string,
+};
+
+export abstract class LensProtocolRouter extends Singleton {
+  public static readonly LoggingPrefix = "[PROTOCOL ROUTER]";
+
+  public abstract on(urlSchema: string, handler: RouteHandler): void;
+  public abstract extensionOn(id: LensExtensionId, urlSchema: string, handler: RouteHandler): void;
+  public abstract removeExtensionHandlers(id: LensExtensionId): void;
+}
+
+/**
+ * This function validates that `options` is at least `BaseParams`
+ * @param args a deserialized value
+ */
+function validateBaseParams(args: unknown): args is BaseParams {
+  if (!args || typeof args !== "object") {
+    // it must be an object
+    return false;
+  }
+
+  if (!hasOwnProperty(args, "handlerType")) {
+    return false;
+  }
+
+  const { handlerType } = args;
+
+  if (handlerType === HandlerType.INTERNAL) {
+    // handlerType must either be HandlerType.INTERNAL
+    return true;
+  }
+
+  if (handlerType === HandlerType.EXTENSION) {
+    if (!hasOwnProperty(args, "extensionId")) {
+      return false;
+    }
+
+    // or handlerType must be HandlerType.EXTENSION
+    const { extensionId } = args;
+
+    // but if for an extension then the extensionId is required, must be a stirng, and must be non-empty
+    return Boolean(extensionId && typeof extensionId === "string");
+  }
+
+  // reject all other values of handlerType
+  return false;
+}
+
+/**
+ * This function validates that `options` is at least `RegisterParams`
+ * @param args a deserialized value
+ */
+export function validateRegisterParams(args: unknown): args is RegisterParams {
+  if (!validateBaseParams(args)) {
+    return false;
+  }
+
+  if (!hasOwnProperties(args, "handlerId", "pathSchema")) {
+    return false;
+  }
+
+  if (typeof args.handlerId !== "string" || args.handlerId.length === 0) {
+    // handlerId is required, must be a string, must be non-empty
+    return false;
+  }
+
+  if (typeof args.pathSchema !== "string" || args.pathSchema.length === 0) {
+    // pathSchema is required, must be a string, must be non-empty
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * This function validates that `args` is at least `DeregisterParams`
+ * @param args a deserialized value
+ */
+export function validateDeregisterParams(args: unknown): args is DeregisterParams {
+  if (args == null || typeof args !== "object") {
```

Comment:
A much safer check would be `if (!args || typeof args !== "object")`

Reply:
How is that safer? `typeof args === "object"` implies either args is `{...}` or `null`.

Reply:
But the preceding `!arg` will catch the `null`, `undefined`, `""`, whatever flasy case.

Note that `typeof [] === 'object'` too, so if you're wanting this to be bulletproof, there's still more to do, but technically, an Array is an object and you can safely get/set any property on it, so I figured that wasn't a concern here.

Technically, `arg == null` should also catch the `undefined` case; it's just so foreign to see a `==` in a JS code base. 😄 Implicit coercion of values on comparison is the source of much trouble...

Reply:
I agree, I like `!args` better than `args == null`. I don't like using the truthy operator `==`. Whenever I see it I wonder if they meant to use `===`.

Reply:
Yeah it isn't a concern to me that `typeof [] === "object"`, also `!args` is still a form of implicit coercion. I'll change it if you guys want though. I just felt that this was more explicit about what I wanted.